File: server/src/skills.ts
import fs from 'fs'
import path from 'path'
import https from 'https'

// Ordered list — persona must come first
const SKILL_FILES = [
  'persona.md',
  'isd-framework.md',
  'playbook.md',
  'decision-making.md',
  'collaboration.md',
  'quality-standards.md',
  'mindsets.md',
]

const CONTENT_DIR = path.resolve(__dirname, '../../src/content')

export function loadSystemPrompt(): string {
  const sections = SKILL_FILES.map((file) => {
    const filePath = path.join(CONTENT_DIR, file)
    if (!fs.existsSync(filePath)) {
      console.warn(`[skills] Missing: ${file} — skipping`)
      return null
    }
    return fs.readFileSync(filePath, 'utf-8').trim()
  }).filter(Boolean)

  return sections.join('\n\n---\n\n')
}

// ── Figma ──────────────────────────────────────────────────────────────────────

// Extracts the file key from any Figma or FigJam URL
// Handles: figma.com/file/KEY, figma.com/board/KEY, figma.com/design/KEY
export function extractFigmaKey(input: string): string | null {
  const match = input.match(/figma\.com\/(?:file|board|design)\/([A-Za-z0-9]+)/)
  return match ? match[1] : null
}

interface FigmaNode {
  id: string
  name: string
  type: string
  children?: FigmaNode[]
  characters?: string // text content on sticky notes etc.
}

interface FigmaFile {
  name: string
  document: FigmaNode
}

function httpsGet(url: string, token: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const options = {
      headers: { 'X-Figma-Token': token },
    }
    https.get(url, options, (res) => {
      let data = ''
      res.on('data', (chunk) => { data += chunk })
      res.on('end', () => resolve(data))
      res.on('error', reject)
    }).on('error', reject)
  })
}

// Recursively walk the node tree and collect a summary
function summariseNode(node: FigmaNode, depth = 0): string[] {
  const lines: string[] = []
  const indent = '  '.repeat(depth)

  // Only include meaningful node types
  const include = ['CANVAS', 'FRAME', 'SECTION', 'GROUP', 'STICKY', 'SHAPE_WITH_TEXT', 'TEXT']
  if (!include.includes(node.type) && depth > 0) return lines

  const label = node.type === 'STICKY' || node.type === 'SHAPE_WITH_TEXT' || node.type === 'TEXT'
    ? `${indent}[${node.type}] "${node.name}"${node.characters ? `: ${node.characters.slice(0, 120).replace(/\n/g, ' ')}` : ''}`
    : `${indent}[${node.type}] ${node.name}`

  lines.push(label)

  // Limit depth to avoid massive output
  if (depth < 3 && node.children) {
    for (const child of node.children.slice(0, 20)) {
      lines.push(...summariseNode(child, depth + 1))
    }
    if (node.children.length > 20) {
      lines.push(`${indent}  ... and ${node.children.length - 20} more`)
    }
  }

  return lines
}

export async function fetchFigmaFile(fileKey: string): Promise<string> {
  const token = process.env.FIGMA_ACCESS_TOKEN
  if (!token) throw new Error('FIGMA_ACCESS_TOKEN not set')

  const url = `https://api.figma.com/v1/files/${fileKey}`
  console.log(`[figma] Fetching file: ${fileKey}`)

  const raw = await httpsGet(url, token)
  const json = JSON.parse(raw) as FigmaFile & { err?: string; status?: number }

  if (json.err || json.status === 403) {
    throw new Error(`Figma API error: ${json.err ?? 'forbidden'}`)
  }

  const lines = [
    `Figma file: "${json.name}"`,
    '',
    ...summariseNode(json.document),
  ]

  return lines.join('\n')
}

File: server/skills.py
import re
import os
from pathlib import Path

import httpx

CONTENT_DIR = Path(__file__).parent.parent / "src" / "content"

CORE_SKILLS = ["persona.md", "isd-framework.md"]

ON_DEMAND_SKILLS = [
    "playbook",
    "decision-making",
    "collaboration",
    "quality-standards",
    "mindsets",
]

_http_client: httpx.AsyncClient | None = None


def _get_http_client() -> httpx.AsyncClient:
    global _http_client
    if _http_client is None:
        _http_client = httpx.AsyncClient()
    return _http_client


def load_core_prompt() -> str:
    sections = []
    for filename in CORE_SKILLS:
        path = CONTENT_DIR / filename
        if not path.exists():
            print(f"[skills] Missing core skill: {filename}")
            continue
        sections.append(path.read_text(encoding="utf-8").strip())
    return "\n\n---\n\n".join(sections)


def load_skill(name: str) -> str:
    path = CONTENT_DIR / f"{name}.md"
    if not path.exists():
        return f'Skill "{name}" not found.'
    print(f"[skills] Loaded on-demand: {name}.md")
    return path.read_text(encoding="utf-8").strip()


def extract_figma_key(url: str) -> str | None:
    match = re.search(r"figma\.com/(?:file|board|design)/([A-Za-z0-9]+)", url)
    return match.group(1) if match else None


_INCLUDE_TYPES = {"CANVAS", "FRAME", "SECTION", "GROUP", "STICKY", "SHAPE_WITH_TEXT", "TEXT"}


def summarise_node(node: dict, depth: int = 0) -> list[str]:
    lines: list[str] = []
    indent = "  " * depth

    if node.get("type") not in _INCLUDE_TYPES and depth > 0:
        return lines

    node_type = node.get("type", "")
    name = node.get("name", "")
    characters = node.get("characters", "")

    if node_type in ("STICKY", "SHAPE_WITH_TEXT", "TEXT"):
        text_preview = characters[:120].replace("\n", " ") if characters else ""
        label = f'{indent}[{node_type}] "{name}"' + (f": {text_preview}" if text_preview else "")
    else:
        label = f"{indent}[{node_type}] {name}"

    lines.append(label)

    if depth < 3:
        children = node.get("children", [])
        for child in children[:20]:
            lines.extend(summarise_node(child, depth + 1))
        if len(children) > 20:
            lines.append(f"{indent}  ... and {len(children) - 20} more")

    return lines


async def fetch_figma_file(file_key: str) -> str:
    token = os.environ.get("FIGMA_ACCESS_TOKEN")
    if not token:
        raise RuntimeError("FIGMA_ACCESS_TOKEN not set")

    url = f"https://api.figma.com/v1/files/{file_key}"
    print(f"[figma] Fetching file: {file_key}")

    client = _get_http_client()
    response = await client.get(url, headers={"X-Figma-Token": token})
    data = response.json()

    if data.get("err") or data.get("status") == 403:
        raise RuntimeError(f"Figma API error: {data.get('err', 'forbidden')}")

    lines = [f'Figma file: "{data["name"]}"', ""]
    lines.extend(summarise_node(data["document"]))
    return "\n".join(lines)
